proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+
+* Uniform Resource Locators (URLs) instructing a client on how to locate a given
+  xDS resource. URLs may include processing directives.
+
+* Uniform Resource Names (URNs) referring to a specific xDS resource. URNs do not
+  include processing directives. `xdstp://` URNs provide self-contained cache keys.
+
+#### Proto3 expression
+
+`xdstp://` URIs are canonically encoded in `proto3` messages. The structure
+for this encoding is provided at https://github.com/cncf/xds/tree/main/xds/core/v3:
+
+* [URL](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_locator.proto) definition
+* [URN](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_name.proto) definition
+
+#### Normalization
+
+Two `xdstp://` URNs are considered equivalent if they match component-wise
+modulo context parameter ordering.
+
+## Context parameters
+
+Context parameters in URNs presented by the client to the server will be dervied
+by the following process. Using an example of a URL
+`xdstp://some-authority/some.type/foo?xds.bar=baz`:
+
+1. Static `Node`-derived context parameters are first added. These are prefixed with
+   `xds.node.`. The set of `Node`-derived context parameters is specified in the
+   bootstrap on a per-resource type basis. The key reflects the `Node` proto3
+   field structure, e.g. `xds.node.locality.sub_zone=some_sub_zone`,
+   `xds.node.metadata.bar="a"`, `xds.node.user_agent_version=1.2.3`.
+
+2. Dynamic context parameters are added next. These are free form and at the
+   discretion of the client. They can provide client-specific fields such as
+   `xds.shard_id=1234` and potentially change through the lifetime of an xDS
+   stream.
+
+3. Context parameters from the URL are added next, in the above example
+   `xds.bar=baz`.
+
+4. [Client feature
+   capability](https://www.envoyproxy.io/docs/envoy/latest/api/client_features)
+   context parameters are then added, e.g. clusters at the client might have
+   `{xds.client_feature.lb.least_loaded: true}`. Client feature capabilities
+   will be required to be declared with respect to a specific resource type URL.
+   All these attributes will be `xds.client_feature.` prefixed.
+
+5. Finally per-resource type well-known attributes are added, e.g. an on-demand
+   listener load might have `{xds.resource.vip: 96.54.3.1}`. These attributes
+   are generated by code, e.g. an on-demand LDS implementation will first detect
+   the VIP, halt the request, ask for the `xds.resource.vip` from the control
+   plane, and then resume the request after discovery is completed. All these
+   attributes will be `xdstp.resource.` prefixed.
+
+An example computed URN following the above example is
+`xdstp://some-authority/some.type/foo?xds.node.metadata.foo=bar&xds.shard_id=1234&xds.bar=baz&xds.client_feature.lb.least_loaded=false&xds.resource.vip=96.54.3.1`.
+
+This proposal reserves the `@` prefix for context parameter values for future
+URI context parameter enhancements.
+
+### Collections
```

Comment:
Do you want to mention which resource type will have list/glob support? e.g. lds, cds.

Reply:
From the perspective of the transport protocol, I don't think there's any limitation here -- there can be collections for any resource type for which that proves useful.

In practice, LDS and CDS are the two main ones today, but we'll also use glob collections for LEDS, and there may be others in the future.  So I don't think we can have an exhaustive list here.

Reply:
Yep, true, I added LDS, CDS and LEDS as non-exhaustive examples.